feat(conditions): Implement on-demand auto-generated conditions.

petervdmeer opened the pull request:
**Feature:** This PR implements the feature request detailed and discussed in issue #5017

## Feature Details
This PR changes the game-conditions (as stored under PlayerInfo, and as used in mission logic) from a `(string)key->(int64_t)value` map to an interface where key->value pairs can still be set, stored and queried, but where it is also possible to register "derived" conditions for querying (and setting) data-values that are stored outside of the PlayerInfo conditions.

A major benefit of this change is that information currently unavailable to content creators (mission writers) can be made available (as int64_t values) by just registering the desired information as derived conditions. Another benefit (as mentioned in issue #5017) is that generated conditions are always "up-to-date", in contrast to being generated and synchronized at predefined moments.

Derived conditions come in 2 flavors; exactly matching single derived conditions (like `flagship crew` or `armament deterrence`) and prefix-matching derived conditions (like `ships: <category>` and `reputation: <government>`).

Derived conditions are not iteratable at the conditions-level, expected usage is that if iteration is needed for data in such derived conditions, that the iteration itself happens on the actual class that contains the data (the players ships-list for `ships: <category>` conditions, and the set of governments for `reputation: <government>` conditions.

All generated conditions are changed to be derived conditions in this PR.

## UI Screenshots
N/A

## Usage Examples
Conditions like `flagship crew`, `armament deterrence`, `ships: Fighter`, as described on the [Wiki](https://github.com/endless-sky/endless-sky/wiki/CreatingMissions#conditions), can just be used by content creators (mission writers) as they were used before, but those conditions are internally no longer stored as conditions, but will be derived from the active flagship and the active players fleet.

## Testing Done
Added relevant automatic unit tests for the following:
- Accounts (test moved to https://github.com/endless-sky/endless-sky/pull/6139)
- ConditionSet
- ConditionsStore

Performed the following tests/checks (and found and fixed 2 bugs with them):
- BankPanel: Checked "Crew Salaries" before and after this change.
- BankPanel: Checked "Salary and Tribute Income" before and after this change.
- OutfitterPanel: Checked in outfitter that licenses were visible.
- Mission: Tested takeoff and land in an active game. Checked job-board.
- Mission: Tested taking a mission and update of offered and done counters.
- Planet: Demanded tribute and checked in savegame that it was set.
- PlayerInfo: Tested by the other tests done.
- PlayerInfoPanel: Checked if the salary/tribute/licenses fields were ok.
- ShipyardPanel: Tested buying a Bactrian (with license: OK).
- ShipyardPanel: Tested buying a Bactrian (without license: OK, blocked).

## Performance Impact
Some processing time will be saved by not always generating all possible automatic conditions at predefined moments. Requesting the same derived conditions multiple times could however be less efficient, since it would be calculated each time it is requested (instead of at the predefined generation moments). We might want to add some caching outside the conditions code to compensate for this if this becomes a problem.

Amazinite commented on source/PlayerInfo.cpp:
```diff
@@ -1721,6 +1721,56 @@ int64_t PlayerInfo::GetCondition(const string &name) const
+
+
+// Remove the value for some condition.
+bool PlayerInfo::EraseCondition(const std::string &name)
```
This file is already `using namespace std`.
```suggestion
bool PlayerInfo::EraseCondition(const string &name)
```

Amazinite commented on source/PlayerInfo.cpp:
```diff
@@ -1721,6 +1721,56 @@ int64_t PlayerInfo::GetCondition(const string &name) const
+
+
+// Get the sum (addition) of all conditions starting with the given prefix.
+int64_t PlayerInfo::GetConditionSum(const std::string &prefix) const
```
```suggestion
int64_t PlayerInfo::GetConditionSum(const string &prefix) const
```

Amazinite commented on source/PlayerInfo.cpp:
```diff
@@ -1721,6 +1721,56 @@ int64_t PlayerInfo::GetCondition(const string &name) const
 
 
+// Copy conditions matching the prefix into the given map.
+void PlayerInfo::GetConditions(std::map<std::string, int64_t> &targetMap, const std::string &prefix) const
```
```suggestion
void PlayerInfo::GetConditions(map<string, int64_t> &targetMap, const string &prefix) const
```

Amazinite commented on source/PlayerInfo.h:
```diff
@@ -180,6 +180,17 @@ class PlayerInfo {
+	int64_t GetConditionSum(const std::string &prefix) const;
+	// Set a "condition" flag to the given value. Returns true on success,
+	// false on failure.
+	bool SetCondition(const std::string &name, int64_t value);
+	bool AddCondition(const std::string &name, int64_t value);
+	bool EraseCondition(const std::string &name);
```
I'm guessing having a return type is setting up for later.

petervdmeer replied:
Yes, that is correct, this is for when we get read-only conditions.

Amazinite commented on source/PlayerInfo.h:
```diff
@@ -180,6 +180,17 @@ class PlayerInfo {
 	
 	// Access the "condition" flags for this player.
 	int64_t GetCondition(const std::string &name) const;
+	// Copy conditions matching the prefix into the given map.
+	void GetConditions(std::map<std::string, int64_t> &targetMap, const std::string &prefix) const;
```
Why modify a map handed to the function instead of returning a map? Are we expecting to have a situation where we want to add to an existing map in a function calling GetConditions?

petervdmeer replied:
Yes, I expect some situations in the future where the map gets filled by recursive calls. Those calls can be implemented by a map that is returned, which will be more readable, or with the current implementation, which should have a somewhat better performance.

The code here is not performance critical, so if we prefer readability, then I can change this to a function that returns a map.

Amazinite replied:
I'll leave it up to tehhowch. I think it's readable enough both ways, but having the function return a map is probably more readable than the current method of modifying a map.

petervdmeer replied:
Will be replaced by iterator (maybe filtered iterator) as per discussion with tehhowch.

tehhowch commented:
Mind putting together a document / some diagrams that show where this is heading? There may be better ways to go about this, such as defining an interface that gets returned by the call to `Conditions()` and having several classes implement that interface, depending on what is supposed to be happening.

petervdmeer commented:
> Mind putting together a document / some diagrams that show where this is heading? There may be better ways to go about this, such as defining an interface that gets returned by the call to `Conditions()` and having several classes implement that interface, depending on what is supposed to be happening.

I just uploaded branch https://github.com/petervdmeer/endless-sky/tree/conditions_provider. That branch is not ready for review/merging yet (it even contains 2 bugs that I already solved in the PR here), but it shows the next step that I want to take by having PlayerInfo implement the new ConditionsProvider interface.

The step after this is to have the automatically generated information (from PlayerInfo::UpdateAutoconditions()) calculated during a call to GetCondition() instead of generated at fixed moments using the UpdateAutoconditions() call. Some of those variables will be read-only, but I can imagine that we make a variable like "unpaid maintenance" also automatically updatable from missions.

If we want to keep the logic for such "unpaid maintenance" in the Accounts class, then we could implement the ConditionsProvider interface on accounts and have PlayerInfo call accounts for all condition-accesses that start with the prefix `"unpaid "`.

The steps after that is to also unlock information in classes like Armament, CargoHold, Engine, Government, Planet, Preferences and Ship for missions by choosing a proper prefix for each piece of information and then implementing the ConditionsProvider interface on those classes.

petervdmeer commented:
> There may be better ways to go about this, such as defining an interface that gets returned by the call to `Conditions()` and having several classes implement that interface, depending on what is supposed to be happening.

When we consider issue #4798 related (as discussed on #5017), then having an object returned that implements the interface seems a better way to go about this. Since returning an object allows for context-specific interfaces.

petervdmeer commented:
Converted to draft; needs Iterator implementation as being added in https://github.com/petervdmeer/endless-sky/tree/conditions_provider before this will be "ready for merge" again.

petervdmeer commented:
Returned to "ready for review" state. PlayerInfo::GetCondition already existed, this PR now only adds PlayerInfo::SetCondition, PlayerInfo::EraseCondition and PlayerInfo::GetConditions. I can take PlayerInfo::GetConditions also out of this PR if it is not good.

petervdmeer commented:
Rebased to resolve merge-conflict.

petervdmeer commented:
I updated this PR with the work to make all accesses indirect and I also added the interfaces for the on-demand conditions to show where this PR is heading to.
Review comments are welcome. Some re-testing needs to be done before we should consider merging this. I hope that the test-framework (#5620 and a few follow-ups) will be done before this PR is ready so that I can create some auto-testers instead of doing manual testing again.

Amazinite commented on source/PlayerInfo.h:
```diff
@@ -189,8 +190,16 @@ class PlayerInfo {
+	// Direct access to "condition" flags data.
+	ConditionsStore &Conditions();
+	const ConditionsStore &Conditions() const;
+	const std::map<std::string, int64_t> &GetManualConditions() const;
```
Just by reading the names, I'm not exactly sure what "manual" is suppose to mean in these method names. Would "raw" be a better term to use? "GetRawConditions" and "EraseRawByPrefix"? Or maybe "direct" for "GetDirectConditions" etc.

petervdmeer replied:
This is the set of conditions that is not On-Demand. I guess `GetNotOnDemandConditions` would also be an appropriate name for this function, but it sounds a bit odd as well.
`Raw` doesn't sound any more clear than `Manual` to me. `Direct` however also okay for me.

petervdmeer replied:
We could also consider if `On Demand` is the best name for conditions that are based on other data, maybe we should rename those to `Derived Conditions`, in which case we can use `GetNonDerivedConditions` as name here (and add a comment that non-derived conditions are stored in a savegame)

Amazinite replied:
Okay, I see where the term comes from. Some conditions are “automatic” (maintained by the source code and writing to them just results in that being overwritten), meaning those conditions which aren’t automatic are manual. 

Are you expecting to have GetAutomaticConditions or similar? If not then maybe it’d be better to just have this be GetConditionsSubset. That way the method name makes it clear that these aren’t all the conditions (or rather, won’t be all the conditions), and then a comment can explain exactly what that subset consists of.

petervdmeer replied:
Let's go for `Derived Conditions` for conditions that are based on other parts of the code and let's use `Primary Conditions` for conditions that are set directly not based on any data in any other class. I'll update the code for this.

The function-name would then be `GetPrimaryConditionsSubset()`.

I don't see a lot of value in getting a list of Derived conditions. The classes that those conditions derive from should already have their own functions to get the relevant data out of those classes.

Amazinite commented on source/PlayerInfo.cpp:
```diff
@@ -1782,36 +1783,88 @@ void PlayerInfo::HandleEvent(const ShipEvent &event, UI *ui)
+// might be okay for a temporary function.
+void PlayerInfo::EraseManualByPrefix(const string &prefix)
+{
+	set<string> toErase;
```
Why can't we erase as we go? Wouldn't you just need to walk back the iterator when something matches and gets erased.

petervdmeer replied:
The list of manual conditions is only provided as `const` (meaning read-only). I could make it writeable as well, but I'm afraid that that could lead to bugs in the future if developers don't notice that they are working with a subset of the conditions.

Amazinite commented on source/ConditionsStore.h:
```diff
@@ -0,0 +1,60 @@
+	ConditionsStore(const std::map<std::string, int64_t> initialConditions);
+
+	// Retrieve a "condition" flag from this provider.
+	int64_t operator [] (const std::string &name) const;
```
Other operator functions don't have spaces around the operator.
```suggestion
	int64_t operator[](const std::string &name) const;
```

Amazinite commented on source/ConditionsStore.cpp:
```diff
@@ -0,0 +1,131 @@
+
+
+
+int64_t ConditionsStore::operator [] (const std::string &name) const
```
```suggestion
int64_t ConditionsStore::operator[](const std::string &name) const
```

Amazinite commented on source/ConditionsStore.cpp:
```diff
@@ -0,0 +1,131 @@
+
+bool ConditionsStore::HasCondition(const std::string &name) const
+{
+	// When we add support for on-demand conditions, then we can
+	// lookup the on-demand condition here before searching the
+	// manually set conditions.
```
Looking forward to this. :)

petervdmeer replied:
Yes, the end-goal of what this PR is preparing for. :-)

Amazinite reviewed (commented):
Looking good so far. I can see the direction that this is headed. Looking forward to what that yields.

petervdmeer commented:
Just updated the ConditionsStore to lookup direct condtions and condition providers in a single to. (This should provide better performance for conditions itself, but it also makes the code is a little bit more complex and it makes accesses to the direct conditions map slower because that map needs to be built on-demand.)

Needs a little bit of re-testing.

petervdmeer commented:
Merged master branch and resolved merge conflicts.

quyykk reviewed (commented):
I don't know what you exactly have in mind for more complex conditions, but I'd like to propose another interface for `ConditionsProvider`.

It seems like every class deriving from `ConditionsProvider` is gonna overwrite the virtual functions in the same exact way but with different values. So, I'd propose having `ConditionsProvider` have 2 virtual functions, maybe named `GetConditionCallbacks` and `SetConditionCallbacks` or something. And what they do is return a map from strings (the condition strings) to a callback. So for `Account` the map for the getter would look like this:

- `"credit"` -> `[this] { return Credits(); }`
- `"net worth"` -> `[this] { return min(limit, max(-limit, NetWorth())); }`

And then `ConditionsProvider` defines all of the functions that `Account` currently overrides.

quyykk commented on source/ConditionsStore.h:
```diff
@@ -0,0 +1,93 @@
+	bool EraseCondition(const std::string &name);
+	
+	// Direct (read-only) access to the stored primary conditions.
+	const std::map<std::string, int64_t> GetPrimaryConditions() const;
```
Good idea! Yeah that will work.

petervdmeer replied:
I uploaded a first version of the iterator. It still needs some work, but the basics seem to be working.

petervdmeer replied:
This method has been removed; all code now uses the iterator instead.

petervdmeer commented on source/Account.cpp:
```diff
@@ -25,6 +25,19 @@ namespace {
+	// Enumaration to minimize the risk of typos compared to using string identifiers.
+	enum class AccountCondition {NET_WORTH, CREDITS, UNPAID_MORTGAGES, UNPAID_FINES, UNPAID_SALARIES, UNPAID_MAINTENANCE, CREDIT_SCORE};
+	
+	const map<AccountCondition, const string> CONDITION_TO_TEXT = {
```
Good ideas, I also considered using static strings in an earlier commit to be able to identify the strings (the enum is not really needed, but I didn't want to copy-paste the strings around either).

petervdmeer replied:
The enum and the map have been removed fully due to the switch to lambda based condition providers, so this review comment is also handled.

petervdmeer commented on source/ConditionsProvider.cpp:
```diff
@@ -0,0 +1,53 @@
+
+// Set a value for a condition. Calls the implementation as provided
+// by the class implementing this interface. Implementing classes are
+// expected to return false and change nothing for read-only conditions.
```
Yes, that is possible, but during an earlier review we considered such defaults vs the benefits of having the need for providers to explicitly specify the desired behavior. If I remember correctly then in that review-session we considered the explicit specification to be more valuable than the reduction in functions to implement. (Providers choose explicitly for read-only, not by accidentally forgetting.)

petervdmeer commented:
> I don't know what you exactly have in mind for more complex conditions, but I'd like to propose another interface for `ConditionsProvider`.
> 
> It seems like every class deriving from `ConditionsProvider` is gonna overwrite the virtual functions in the same exact way but with different values. So, I'd propose having `ConditionsProvider` have 2 virtual functions, maybe named `GetConditionCallbacks` and `SetConditionCallbacks` or something. And what they do is return a map from strings (the condition strings) to a callback. So for `Account` the map for the getter would look like this:
> 
>     * `"credit"` -> `[this] { return Credits(); }`
>     * `"net worth"` -> `[this] { return min(limit, max(-limit, NetWorth())); }`
> 
> 
> And then `ConditionsProvider` defines all of the functions that `Account` currently overrides.

Using lambda functions/callback functions would indeed be very elegant. I did some tests with them in branch https://github.com/petervdmeer/endless-sky/commits/conditions_provider. Storing the lambdas/callbacks is however a bit tricky in C++11 and I didn't get it fully working yet.

quyykk commented:
> Storing the lambdas/callbacks is however a bit tricky in C++11 and I didn't get it fully working yet.

How so? `std::function`? Since it the lambdas would only capture the `this` pointer, it should be well within the real of SOO thus incurring minimal impact on performance compared to a regular function pointer :)

petervdmeer commented:
> > Storing the lambdas/callbacks is however a bit tricky in C++11 and I didn't get it fully working yet.
> 
> How so? `std::function`? Since it the lambdas would only capture the `this` pointer, it should be well within the real of SOO thus incurring minimal impact on performance compared to a regular function pointer :)

Ah, looks like I missed the difference between `std::function` and function pointers in my other branch.
If it is indeed possible to convert lambdas (with capture sections) to `std::function` and then store them in a map or vector, then that could also lead to better performance (since the ConditionsStore could lookup a relevant lambda directly in the internal map and execute it without further lockups, the current implementation has a second lookup at Account.h which would disappear).
This is probably worth looking into (but I'm quite busy this weekend, so it might be for later in August).

petervdmeer commented:
Converted to draft; branch https://github.com/petervdmeer/endless-sky/tree/conditions_provider now contains a (mostly working) implementation where derived-conditions are retrieved using lambda functions (I finally got that working thanks to quyykk s remark on the use of `std::function`). I would like to transfer that implementation into this PR since I expect that implementation to work much better than the inheritance currently done with ConditionsProvider.

quyykk commented on source/ConditionsStore.cpp:
```diff
@@ -0,0 +1,276 @@
+
+
+
+const ConditionsStore::ConditionEntry *ConditionsStore::GetEntryConst(const string &name) const
```
Why not just `GetEntry`? Why the extra `Const` in the name?

petervdmeer replied:
Agreed, the extra Const in the name is probably not needed, I'll check if I can remove it.

quyykk replied:
Also, since the 2 functions are the same I'd consider using the `const_cast` hack if it doesn't violate the style guide :)

quyykk commented on source/ConditionsStore.h:
```diff
@@ -0,0 +1,152 @@
+
+private:
+	// Storage for both the primary conditions as well as the providers.
+	std::map<std::string, ConditionEntry> storage;
```
Quick question: Have you considered having 3 maps instead of one? It might simplify lots of code (like the iterator not being needed anymore). We'll need to do 3 map lookups in the worst case to get a condition though but map lookups aren't expensive. I haven't fully thought it out so just a thought.

petervdmeer replied:
Yes, earlier implementations had 2 or 3 separate maps, but tehhowch was concerned about the performance when we would need to do all those separate lookups, so I switched to this (more complex) implementation that gets the relevant condition(provider) in a single lookup.

quyykk replied:
I guess he has a point. Although IMO: The majority of condition access are for value conditions right? Sooo you'd only have multiple lookups for anything other than a value condition....... :)

petervdmeer replied:
Derived conditions take precedence over value conditions, so you would have 2 look-ups for value conditions:
-1- Check if a derived condition exists with the given name or prefix.
-2- Get the relevant value condition.

I guess we could do something smart to actively prune derived conditions out of the value conditions, but even then we would need 2 lookups for an unset value condition:
-1- Check if the value exists in the list of value conditions.
-2- If it is not set, then check if a derived condition exists.

quyykk commented on source/ConditionsStore.h:
```diff
@@ -0,0 +1,152 @@
+
+private:
+	// Retrieve a condition entry based on a condition name, doCreate indicate if the entry
+	// should be created if it wasn't there yet.
```
Wahts `doCreate` that this comment alludes to?

petervdmeer replied:
An artifact of some old code (where the has-function should never create new entries, while the setFun should). I think I can completely remove this comment now.

quyykk commented on source/ConditionsStore.h:
```diff
@@ -0,0 +1,152 @@
+		std::function<bool(const std::string&)> hasFun;
+		std::function<bool(const std::string&, int64_t)> setFun;
+		std::function<bool(const std::string&)> eraseFun;
+	};
```
What would be a condition where `hasFun` and `eraseFun` would do something different than the default implementation? Especially `eraseFun`. Can't really delete a auto generated condition right?

petervdmeer replied:
I expect nearly all auto-generated conditions to be read-only, but suppose that we make mortgages a read-write auto-generated condition, then erasing the condition that refers to a specific mortgage could remove that specific mortgage. (Not a very good example, because I expect the mortgages to be provided read-only if we make them available through conditions.)

petervdmeer replied:
The [tributes](https://github.com/endless-sky/endless-sky/issues/5448) might be a better example.

petervdmeer commented:
All automatic conditions are now implemented as derived conditions, but not all of them were tested yet. (Some are covered by automatic tests, but not all of them yet.)

Amazinite reviewed (changes requested):
I will admit that some of the stuff in ConditionStore is definitely outside of my knowledge base, but from what I can tell everything looks good and checks out. My only comments are on a few minor errors.

I've also opened a PR to this branch with some style enforcement changes: https://github.com/petervdmeer/endless-sky/pull/8

petervdmeer commented on source/PlayerInfo.cpp:
```diff
@@ -2520,76 +2480,157 @@ void PlayerInfo::ValidateLoad()
+		int64_t retVal = 0;
+		for(const shared_ptr<Ship> &ship : ships)
+			if(!ship->IsParked() && !ship->IsDisabled() && ship->GetSystem() == system
+					&& name == "ship: " + ship->Attributes().Category())
```
```suggestion
					&& name == "ships: " + ship->Attributes().Category())
```

petervdmeer commented on source/PlayerInfo.cpp:
```diff
@@ -2520,76 +2480,149 @@ void PlayerInfo::ValidateLoad()
+				++retVal;
+		return retVal;
+	};
+	conditions.SetProviderPrefixed("ships: ", conditionsProvider);
```
No, that was a bug. Should be `"ships: "` on both values => changed this.

petervdmeer commented on source/PlayerInfo.cpp:
```diff
@@ -2520,76 +2480,149 @@ void PlayerInfo::ValidateLoad()
-	}
+	conditionsProvider.getFun = [this] (const string &name)->int64_t {
+		if(flagship && !boardingMissions.empty())
+			flagship->Cargo().Free();
```
Yes, the return was missing.

petervdmeer commented:
Rebased to solve merge-conflict.

petervdmeer commented:
Merged master to resolve merge-conflict.

petervdmeer commented:
Converted to draft, want to try to remove the PrimariesInterator completely:
- Moving the saving (serialization) to ConditionsStore will remove the iterator for saving.
- Allowing providers to save through ConditionsStore (through a callback) will also allow splitting off iteraterable items (removing the need to centrally iterate).

Amazinite commented:
@petervdmeer Is there anything that I or anyone else could do to help move this along, or is it mainly just waiting on you refactoring it? Like #5440, this is one of those PRs that repeatedly gets referenced as something that could greatly improve current mission writing capabilities, so I'd like to get this finalized and merged soon.

Amazinite commented on source/ConditionsStore.cpp:
```diff
@@ -0,0 +1,456 @@
+
+
+
+ConditionsStore::ConditionEntry &ConditionsStore::operator[](const std::string &name)
```
Style change: the file is already using namespace std. There are various other instances of `std::` in this file, but those are on standard function calls, in which case we typically do prepend `std::` to those.
```suggestion
ConditionsStore::ConditionEntry &ConditionsStore::operator[](const string &name)
```

Amazinite commented on source/ConditionsStore.cpp:
```diff
@@ -0,0 +1,456 @@
+
+
+
+void ConditionsStore::DerivedProvider::SetGetFun(std::function<int64_t(const std::string&)> newGetFun)
```
Judging by other files, the style is to drop the `std::` from here too.
https://github.com/endless-sky/endless-sky/blob/9124602fe322f4ef92ec920a3ec89ca7bd49022a/source/Panel.cpp#L72
```suggestion
void ConditionsStore::DerivedProvider::SetGetFun(function<int64_t(const string&)> newGetFun)
```
There are three more functions below this one that would also need this change.

petervdmeer replied:
This review-comment is not on the latest commit; this was already fixed.

Amazinite commented on source/ConditionsStore.h:
```diff
@@ -0,0 +1,197 @@
+
+		// Default input_iterator operations.
+		std::pair<std::string, int64_t> operator*() const;
+		const std::pair<std::string, int64_t>* operator->();
+		PrimariesIterator& operator++();
+		PrimariesIterator operator++(int);
+		bool operator== (const PrimariesIterator& rhs) const;
+		bool operator!= (const PrimariesIterator& rhs) const;
```
Style change:
```suggestion
		const std::pair<std::string, int64_t> *operator->();
		PrimariesIterator &operator++();
		PrimariesIterator operator++(int);
		bool operator==(const PrimariesIterator &rhs) const;
		bool operator!=(const PrimariesIterator &rhs) const;
```

petervdmeer replied:
Style fix was done by patch from samrocketman.

Amazinite reviewed (dismissed):
Reasserting my approval since it's been a while since my last review. I only have a few style-related comments, but outside of that—from what I can understand of the ConditionsStore—this looks like some great work.

Hurleveur commented on source/ConditionSet.cpp:
```diff
@@ -162,12 +163,12 @@ namespace {
-					value = temp->second;
-				else if(perm != conditions.end())
-					value = perm->second;
+				bool tempHas = created.Has(str);
+				bool permHas = conditions.Has(str);
+				if(tempHas)
+					value = created.Get(str);
+				else if(permHas)
+					value = conditions.Get(str);
```
Is this an optimization? Idk how exactly has and find work, but we could be looking for it twice when it's there?

petervdmeer replied:
You're right, this code is indeed looking for the conditions twice when they are there. I'll have a look if I can improve on this.

petervdmeer replied:
I added an `HasGet` function which performs the Has and Get functions with a single lookup (and returns a pair with the results).

An alternative approach would be to implement a Find which returns a `ConditionEntry`, but the ConditionEntry will not always exist for derived conditions, and I believe it is performance-wise better not to create it when we are only performing read-access.

Hurleveur replied:
Looks good to me this way

Hurleveur commented on source/PlayerInfo.cpp:
```diff
@@ -2564,76 +2519,151 @@ void PlayerInfo::ValidateLoad()
+			return 0;
+		return gov->Reputation();
+	});
+	reputationProvider.SetSetFun([](const string &name, int64_t value) -> bool
```
Just wondering, why is this the only one that has a setter?
reputation is the only primary condition, that is not generated by code and is written in the save file, right?

maybe it could be explained a comment

petervdmeer replied:
Reputations with governments are stored in `GameData::Governments()` and they typically get copied between the conditions and the governments. This PR removes all the 2-directional copying and just performs reads and writes of those conditions directly in the `GameData::Governments()` reputation data.
This also means that reputations are not primary conditions (according to terminology in this PR they are derived conditions), and the reputation conditions will no longer be stored in savefiles, but the Government reputations still are.

Hurleveur commented on source/Test.cpp:
```diff
@@ -560,13 +560,13 @@ void Test::Fail(const TestContext &context, const PlayerInfo &player, const stri
 	if(!conditions.empty())
 		Files::LogError(conditions);
-	else if(player.Conditions().empty())
+	else if(player.Conditions().PrimariesBegin() == player.Conditions().PrimariesEnd())
```
You already have an empty() function defined, does this not check for the same thing?
Btw does it only checks for primaries? Probably should be clear in the function's name.

petervdmeer replied:
The `ConditionsStore::empty()` function does not have a well-defined meaning when we are working with derived conditions. I removed this function.

Hurleveur reviewed (commented):
Is it a convention to call the functions Fun? (~~we having fun coding I guess~~)
If not maybe it would be clearer replacing that by Function, we are defining custom getters, setters etc after all.

Looks pretty solid, finally we no longer need to stock conditions as separate from the actual thing.
Indeed the tests working means its all good. Nice that you've added some more for this.
There wasn't much freedom of action in this; what I mean is that most things had to be changed this way. Still I tried to comment on the parts I think can be changed.
Code wise it's much more elegant this way.

samrocketman reviewed (commented):
I was able to do a style review.  I would like to do a "part 2" where I exercise the game logic.  However, I think there's value in submitting the style review first in case it is desirable for this to be merged before I get a chance to more thoroughly review the code.  As well as play test the code.

# What was tested

Ran unit tests.
Ran integration tests.
Started a new game with sounds of music and the music started playing after leaving the planet.

From this initial play test the conditions haven't drastically changed in a negative way.  I look forward to testing out the dynamic nature of the conditions this feature provides.

# Code style fixes

I have summarized all of your code style issues into a convenient patch that you can apply as opposed to pointing everything out in this PR.

How to apply:

1. [Download the patch](https://gist.github.com/samrocketman/5d94bb1f6d2ff06edaa4b65d3e274e17).
2. Run the command `git apply style-changes.patch`.

The style fixes are based on commit dc7f347a0f00495183f394cbd964c9a8cdd29b26

petervdmeer commented on tests/unit/src/test_conditionsStore.cpp:
```diff
@@ -0,0 +1,579 @@
+
+// #region mock data
+
+void verifyName(const std::string &name1, const std::string &name2)
```
I compared the indentation of those tests with the other tests; all other tests also didn't have the indentation as suggested. If we want to fix this, then we should fix it for all tests at once. I applied your patch, but removed this additional indentation for this specific test.

Maybe we should also fix this by some specific entries in the style guide about the indentation specific for the catch-based unit-tests; I feel that avoiding the initial indentation in the catch-specific files actually increases readability.

petervdmeer commented on source/ConditionsStore.h:
```diff
@@ -0,0 +1,197 @@
+	// Class for DerivedProviders, the (lambda) functions that provide access
+	// to the derived conditions are registered in this class.
+	class DerivedProvider
+	{
```
Thanks for the patch, I applied it.

Hurleveur reviewed (approved):
Seems good to me now, this is a very good improvement to the quality of the code in my opinion.

samrocketman commented on tests/unit/src/test_conditionsStore.cpp:
```diff
@@ -0,0 +1,702 @@
+	{
+		auto store = ConditionsStore();
+		auto mockProvPrefixShips = MockConditionsProvider();
+		mockProvPrefixShips.SetRWPrefixProvider(store, "ships: " );
```
```suggestion
		mockProvPrefixShips.SetRWPrefixProvider(store, "ships: ");
```

My mistake on this

quyykk commented on tests/unit/src/test_conditionSet.cpp:
```diff
@@ -28,6 +31,19 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+
+int primarySize(const ConditionsStore &store)
+{
+	int size = 0;
+	auto it = store.PrimariesBegin();
+	while(it != store.PrimariesEnd())
+	{
+		++it;
+		++size;
+	}
+	return size;
```
You can use `std::distance` here.

petervdmeer replied:
I added a TODO item for this. (As you noticed before; the iterator that is needed for this is not yet correctly implemented.)

quyykk reviewed (commented):
Nice, I think this looks good now. If you can the primary iterators in a later refactor that would be great yeah.

petervdmeer commented on source/ConditionsStore.h:
```diff
@@ -0,0 +1,187 @@
+		PrimariesIterator& operator++();
+		PrimariesIterator operator++(int);
+		bool operator== (const PrimariesIterator& rhs) const;
+		bool operator!= (const PrimariesIterator& rhs) const;
```
This review-comment is not on the latest commit; this was already fixed.

petervdmeer commented on source/PlayerInfo.cpp:
```diff
@@ -2564,76 +2519,151 @@ void PlayerInfo::ValidateLoad()
+
+	// Read-only flagship conditions.
+	auto &&flagshipCrewProvider = conditions.GetProviderNamed("flagship crew");
+	flagshipCrewProvider.SetGetFun([this](const string &name) -> int64_t { if(flagship) { return flagship->Crew(); } return 0; });
```
Yes, that's nice. I updated the code as suggested.

petervdmeer commented on source/ConditionsStore.h:
```diff
@@ -0,0 +1,187 @@
+		using value_type = std::pair<const std::string, int64_t>;
+		using difference_type = void;
+		using pointer = const std::pair<const std::string, int64_t>*;
+		using reference = std::pair<const std::string, int64_t>;
```
This code is indeed not correct as it is currently implemented. I expect that we can fix this by returning pointers to ConditionEntries instead of the pairs that are used now, but for now I added some TODO texts to the code here to at least signal the issues. I also started [a new PR](https://github.com/endless-sky/endless-sky/pull/7075) to remove the last condition-iterators, after which the PrimariesIterator can also be fully removed.

I am a bit affraid that tehhowch might care. He has some ideas on conditions of which I'm not fully aware yet, and he did remark in the past that he saw a need to be able to iterate over all possible conditions that exist. I'm not sure if this remark still applies after this PR and after #7075.

Hurleveur reviewed (approved):
Do I have to click viewed on every single file for it to not dismiss my review?
In general I don't do that because I like to look back and forth between the classes.

tehhowch reviewed (commented):
Sharing whatever comments I had stashed for when I was able to review.

(App doesn't want to make seeing which ones are relevant still easy to see)

petervdmeer commented on source/PlayerInfo.cpp:
```diff
@@ -1721,6 +1721,43 @@ int64_t PlayerInfo::GetCondition(const string &name) const
 
 
+// Copy conditions matching the prefix into the given map.
+void PlayerInfo::GetConditions(map<string, int64_t> &targetMap, const string &prefix) const
```
This code is no longer in use. The latest version of the code returns a ConditionsStore, which seems to be more elegant than this code.

petervdmeer commented on source/ConditionsProvider.h:
```diff
@@ -0,0 +1,54 @@
+class ConditionsProvider {
+public:
+	// Retrieve a "condition" flag from this provider.
+	virtual int64_t GetCondition(const std::string &name) const = 0;
```
Agreed, but this code is also no longer in use; the Provider interfaces are now defined as a subclass within ConditionsStore, and the abstraction is done using lambda functions..

petervdmeer commented on source/ConditionsStore.h:
```diff
@@ -0,0 +1,67 @@
+// provider or as basis for other classes that need the same implementation
+// as is done in this class.
+// Classes that want to store conditions differently than that is done in
+// this class should inherit from ConditionsProvider.
```
No classes derive from the providers anymore in the latest version of the code.

petervdmeer commented on source/SpaceportPanel.cpp:
```diff
@@ -112,7 +112,7 @@ const News *SpaceportPanel::PickNews() const
 	vector<const News *> matches;
 	const Planet *planet = player.GetPlanet();
-	const map<string, int64_t> &conditions = player.Conditions();
+	auto &conditions = player.Conditions();
```
Good comment, [quyykk also noticed this](https://github.com/endless-sky/endless-sky/pull/6373#discussion_r776866023) and the issue was already fixed.

petervdmeer commented on source/ConditionsStore.h:
```diff
@@ -0,0 +1,152 @@
+	//         are all handled by the given provider.
+	// - EXACT_PROVIDER means that the condition with the exact name is handled
+	//         by the given provider.
+	enum StorageType {VALUE, PREFIX_PROVIDER, EXACT_PROVIDER};
```
Good comment, but the enum no longer exists in the latest version of the code.

petervdmeer commented on source/ConditionsStore.h:
```diff
@@ -0,0 +1,152 @@
+	typedef std::function<int64_t(const std::string&)> GetFun;
+	typedef std::function<bool(const std::string&)> HasFun;
+	typedef std::function<bool(const std::string&, int64_t)> SetFun;
+	typedef std::function<bool(const std::string&)> EraseFun;
```
The latest version of the code uses private `std::function` variables instead of the typedefs.

petervdmeer commented on source/ConditionsStore.cpp:
```diff
@@ -0,0 +1,276 @@
+		return &(it->second);
+	
+	return nullptr;
+}
```
Already fixed based on [the same comment from quyykk](https://github.com/endless-sky/endless-sky/pull/5244#discussion_r687091461).

petervdmeer commented on source/ConditionsStore.cpp:
```diff
@@ -0,0 +1,276 @@
+void ConditionsStore::PrimariesIterator::MoveToValueCondition()
+{
+	while((condMapIt != condMapEnd) && (condMapIt->second).type != StorageType::VALUE)
+		condMapIt++;
```
> If you're incrementing `condMapIt` every time before calling this

We don't increment before the call if/when we create the initial iterator. (the very first position could already contain a valid value.)

petervdmeer commented on source/ConditionsStore.cpp:
```diff
@@ -0,0 +1,276 @@
+	while((condMapIt != condMapEnd) && (condMapIt->second).type != StorageType::VALUE)
+		condMapIt++;
+	
+	if(condMapIt != condMapEnd)
+		itVal = make_pair(condMapIt->first, (condMapIt->second).value);
+}
```
I'll add a comment (in https://github.com/endless-sky/endless-sky/pull/7226)

petervdmeer commented on source/ConditionSet.cpp:
```diff
@@ -162,12 +163,12 @@ namespace {
+				if(tempHas)
+					value = created.GetCondition(str);
+				else if(permHas)
+					value = conditions.GetCondition(str);
```
The permHas is indeed moved within the else part (since we don't need to look up anything for it if the tempHas matches).

petervdmeer commented on source/ConditionsStore.h:
```diff
@@ -0,0 +1,151 @@
+	// Retrieve a condition entry based on a condition name, doCreate indicate if the entry
+	// should be created if it wasn't there yet.
+	ConditionEntry *GetEntry(const std::string &name);
+	const ConditionEntry *GetEntryConst(const std::string &name) const;
```
I feel it's more elegant to explicitly handle this at the functions themselves; having the casts at only a single location.

petervdmeer commented on tests/src/test_conditionsStore.cpp:
```diff
@@ -0,0 +1,498 @@
+			THEN( "not given conditions return the default value" ) {
+				REQUIRE( 0 == store.GetCondition("ungreeted world") );
+				REQUIRE( primarySize(store) == 2 );
+				REQUIRE( store.GetCondition("ungreeted world") == 0 );
```
Agreed, this is a bit overkill for the Get function. I'll improve this code in https://github.com/endless-sky/endless-sky/pull/7226.

petervdmeer commented on tests/src/test_conditionsStore.cpp:
```diff
@@ -0,0 +1,498 @@
+		auto store = ConditionsStore();
+		REQUIRE( primarySize(store) == 0 );
+		WHEN( "a condition is set" ) {
+			REQUIRE( store.SetCondition("myFirstVar", 10) == true );
```
Good point; I will improve this in https://github.com/endless-sky/endless-sky/pull/7226

petervdmeer commented on source/ConditionSet.cpp:
```diff
@@ -546,21 +547,21 @@ bool ConditionSet::Expression::Test(const Conditions &conditions, const Conditio
+	int64_t c = created.GetCondition(Name());
 	int64_t value = right.Evaluate(conditions, created);
-	c = fun(c, value);
+	created.SetCondition(Name(), fun(c, value));
```
Was fixed in a different way by adding `operator[]`.
